<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>DOMINO by molevol-ub</title>
  </head>

  <body>
    <header>
      <div class="inner">
      <p> <a href="http://www.ub.edu/softevol/domino/">
      		<img src="images/DOMINO_icon-small.png" title="DOMINO website" style="float:left;border:50px;width:120px;heigh=191px">
      </a></p>

      <h1>DOMINO</h1>
      <h2>Development Of Molecular Markers In Non-Model Organisms</h2>
      <a href="https://github.com/molevol-ub/DOMINO" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a id="1-introduction" class="anchor" href="#1-introduction" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>1. INTRODUCTION</h1>

<p>DOMINO is a new software application specifically designed for improving the development of DNA markers in non-model organisms using either NGS data or pre-computed multiple sequence alignments (MSA) in various formats (including RAD loci). We used Perl and C++ scripting languages to combine some popular, open-source, bioinformatics tools for NGS data with a set of new developed functions in an integrated bioinformatics pipeline for custom-made marker discovery or selection based on user-defined criteria.</p>

<p>Customizable features include the length of variable and conserved regions (when requested), the minimum levels (or a preferred range) of nucleotide variation, how to manage polymorphic variants, or which taxa (or what fraction of them) should be covered by the marker. All these criteria can be easily defined in a user-friendly GUI or under a command-line version that implements some extended options and that it is particularly useful for working with large NGS data sets in high performance computers (see also the DOMINO documentation). The regions identified or selected in DOMINO can be i) directly used as markers with a particular depth of taxonomic resolution, ii) utilized for their downstream PCR amplification in a broader taxonomic scope or iii) used as suitable templates to optimized bait design for target DNA enrichment techniques.</p>

<p>The DOMINO workflow consists in four main phases: “Input Data” and “Pre-processing”, “Assembly”, “Alignment/Mapping” and “Marker Discovery/Selection”. Users can enter to the DOMINO workflow at different points in order to perform specific shortcut runs for marker discovery or selection according to the type of input data. In a typical full run from raw NGS reads (using marker identification module), the program applies an assembly-based approach; the pipeline is therefore optimized to work with genome partitioning methods in which the size-selected or enriched fragments are long enough to permit a successful assembly and have been fully (or nearly fully) sequenced. For MSA inputs, as for example, restriction-site associated DNA (RAD) variation and related methods (e.g. RADseq, ddRAD, GBS) DOMINO can search the MSA of each loci (in this case RAD loci) previously obtained with commonly used software packages for highly informative markers (marker selection module); current version accepts the PHYLIP and FASTA format as well as the pyRAD (.loci) and STACKS (.fa) output file formats.</p>

<p>After the marker development phase, DOMINO provides i) a list of the genomic regions (with their corresponding coordinates in each generated contig) that meet marker design criteria and the MSA of each of these markers MSAs for the specified taxa (DOMINO marker identification module outputs) or ii) the list of makers selected under the same criteria among a set of aligned sequences supplied by the user and previously obtained in other software (DOMINO marker selection module outputs).</p>

<h1>
<a id="2-documentation" class="anchor" href="#2-documentation" aria-hidden="true"><span aria-hidden="true" class="times new roman"></span></a> 2. DOCUMENTATION </h1>

<p>Documentation and additional information can be retrieved from the DOMINO oficial website: <a href="http://www.ub.edu/softevol/domino/">http://www.ub.edu/softevol/domino</a></p>

<h1>
<a id="2-installation" class="anchor" href="#2-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3. INSTALLATION</h1>

<h2>
<a id="21-supported-operating-systems" class="anchor" href="#21-supported-operating-systems" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3.1. Supported Operating Systems</h2>

<p>DOMINO is developed and maintained under GNU/Unix using ANSI/ISO C and standard development tools (GCC, Autotools).</p>

<p>It can be compiled and run in Linux (it has been tested in Ubuntu 14.04 LTS) and Mac OSX systems (version &gt; 10.10)</p>

<h2>
<a id="22-requirements" class="anchor" href="#22-requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3.2. Requirements</h2>

<ul>
<li><p>Perl Programming Language version v5.2+</p></li>
<li>
<p>C++ compiler</p>
<ul>
<li>Mac OS users may need to install Xcode package via App Store</li>
<li>Linux users may need to install build essential package o g++ compiler
Ubuntu:
    sudo apt-get install build-essential 
    or 
    sudo apt-get install g++</li>
</ul>
</li>
<li>
<p>zlib compression library v1.2.3+ (SAMtools dependency) <a href="http://www.zlib.net/">http://www.zlib.net/</a></p>
<ul>
<li>Mac OS: included and installed in the distribution.</li>
<li>Linux users may need to install zlib1g-dev package Ubuntu/Linux Mint:
    sudo apt-get install zlib1g-dev</li>
</ul>
</li>
<li>
<p>Xterm (in some ubuntu-like distributions may not be included):</p>
<ul>
<li>Mac OS: included and installed in the distribution</li>
<li>Ubuntu/Linux Mint:
sudo apt-get install xterm</li>
</ul>
</li>
</ul>

<h2>
<a id="23-installation" class="anchor" href="#23-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3.3. Installation</h2>

<p>Two possible installation options:</p>

<h3>
<a id="231-domino-gui" class="anchor" href="#231-domino-gui-plus-command-line-version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3.3.1. DOMINO GUI:</h3>

<p>In the DOMINO website, user can obtain installers that will guide the user through the complete installation process for Linux and Mac distributions.</p>

<p>Also, all the necessary installation files and folders (the user may need to compile and retrieve some Qt libraries) are available at the domino Git repository.</p>

<p>Once installed, the bin directory will contain the DOMINO perl scripts, the directories with the binaries of the external software included in the DOMINO pipeline and some mandatory Perl modules. In addition, a shortcut Desktop DOMINO icon will be generated.</p>

<h3>
<a id="232-command-line" class="anchor" href="#232-command-line" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3.3.2. Command-line:</h3>

<p>In the Git directory, the user can find a pre-configured shell script (install.sh) written to handle the installation of command-line version (to install only the command-line version).</p>

<p>In a Unix-based bash console, change the directory to enter in the desired installation folder and run the script, with the command: sh install.sh</p>

<p>The bin directory will contain the DOMINO perl scripts, the directories with the binaries of the external software included in the DOMINO pipeline and some mandatory Perl modules.</p>

<h1>
<a id="2-bug-reports" class="anchor" href="#2-bug-reports" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>4. BUG REPORTS</h1>

<p>If you encounter bugs that are not listed here, please send comments and bug reports via GitHub</p>

<h1>
<a id="3-community" class="anchor" href="#3-community" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>5. COMMUNITY</h1>

<p>If you would like to see a new feature implemented, suggestions are welcome. Our aim is to have a "suggestion" box soon.</p>

<h1>
<a id="4-thanks" class="anchor" href="#4-thanks" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>6. THANKS</h1>

<p>The authors would like to thank all people who contributed (and are still contributing) in the creation and testing of this software. </p>

<h1>
<a id="5-authors" class="anchor" href="#5-authors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>7. AUTHORS</h1>

<p>Cristina Frias-Lopez, Jose F. Sanchez-Herrero, Miquel A. Arnedo, Alejandro Sanchez-Gracia and Julio Rozas.</p>
<pre><code>C.Frias-Lopez: cristinafriaslopez@ub.edu
J.F.Sanchez: jfsanchezherrero@ub.edu
M.A. Arnedo: marnedo@ub.edu
A.Sanchez-Gracia: elsanchez@ub.edu
J.Rozas: jrozas@ub.edu 
</code></pre>

<p>Evolutionary Genomics and Bioinformatics Group, Departament de Genètica, Microbiologia i Estadística and Institut de Recerca de la Biodiversitat (IRBio), Universitat de Barcelona, Av. Diagonal 643, Barcelona 08028, Spain</p>

<h1>
<a id="6-citation" class="anchor" href="#6-citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>8. CITATION</h1>

<p><em>Bioinformatics first published online August 16, 2016 <a href="http://bioinformatics.oxfordjournals.org/content/early/2016/09/02/bioinformatics.btw534.full"> doi:10.1093/bioinformatics/btw534</a></em></p>


<h1>
<a id="7-copyright-and-license" class="anchor" href="#7-copyright-and-license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>9. COPYRIGHT AND LICENSE</h1>

<p>Copyright (C) 2016 Evolutionary Genomics and Bioinformatics Group, University of Barcelona.</p>

<p>DOMINO is licensed under the GPLv3 license.  See 'LICENSE' file for details. </p>

<p>This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.</p>

<p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  </p>

<p>See the GNU General Public License for more details.</p>

<p>You should have received a copy of the GNU General Public License along with this program.<br>
If not, see <a href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a></p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/molevol-ub/DOMINO/zipball/master" class="button" title="Source code latest release (.zip)">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/molevol-ub/DOMINO/tarball/master" class="button" title="Source code latest release (.tar.gz)">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/molevol-ub/DOMINO"></a> DOMINO is maintained by <a href="https://github.com/molevol-ub">molevol-ub</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>
  
  </body>
</html>
