<!DOCTYPE html>
<html>
  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
    <link href='https://fonts.googleapis.com/css?family=Architects+Daughter' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/print.css" media="print">

    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <title>DOMINO by molevol-ub</title>
  </head>

  <body>
    <header>
      <div class="inner">
        <h1>DOMINO</h1>
        <h2>Development Of Molecular Markers In Non-Model Organisms</h2>
        <a href="https://github.com/molevol-ub/DOMINO" class="button"><small>View project on</small> GitHub</a>
      </div>
    </header>

    <div id="content-wrapper">
      <div class="inner clearfix">
        <section id="main-content">
          <h1>
<a id="1-introduction" class="anchor" href="#1-introduction" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>1. INTRODUCTION</h1>

<p>DOMINO is a new software application specifically designed for improving the development of
DNA markers in non-model organisms using either NGS data or pre-computed multiple sequence 
alignments (MSA) in various formats (including MSA from RAD data). </p>

<p>We used Perl and C++ scripting languages to combine some popular, open-source, bioinformatics
tools for NGS data with a set of new developed functions in an integrated bioinformatics
pipeline for custom-made marker discovery or selection. </p>

<p>Using NGS data (raw or pre-processed reads) from typical genome partitioning or low depth
sequencing experiments (including or not reference sequences, such as, for example, genome
scaffolds), or pre-computed multiple sequence alignments (MSA) of multiple taxa 
(“taxa panel”, which may represent different populations or species), DOMINO searches for 
sequence regions with a minimum level of variation, which can be optionally flanked by 
stretches of conserved sequences (appropriate for further PCR primers design). </p>

<p>The DOMINO workflow consists in four main phases: “Input Data” and “Pre-processing”, 
“Assembly”, “Alignment/Mapping” and “Marker Discovery/Selection”. Users can enter to the 
DOMINO workflow at different points in order to perform specific shortcut runs for marker 
discovery or selection according to the type of input data.</p>

<p>In a typical full run from raw NGS reads (using marker identification and marker selection 
modules), the program applies an assembly-based approach; the pipeline is therefore 
optimized to work with genome partitioning methods in which the size-selected or enriched 
fragments are long enough to permit a successful assembly and have been fully (or nearly 
fully) sequenced. For MSA inputs, as for example, restriction-site associated DNA (RAD) 
variation and related methods (e.g. RADseq, ddRAD, GBS) DOMINO can searh the MSA of each 
loci (in this case RAD loci) previously obtained with commonly used software packages for 
highly informative markers (marker selection module); current version accepts the PHYLIP 
and FASTA format as well as the pyRAD (<em>.loci) and STACKS (</em>.fa) output file formats. </p>

<p>After the maker development phase, DOMINO provides i) the list the genomic regions (with 
their corresponding coordinates in each contig) exhibiting the minimum levels of nucleotide 
variation specified by the user among the selected taxa (or covering at least a pre-defined 
minimum number of them), and the marker MSAs for these taxa (DOMINO marker identification 
module) or ii) the list of markers exhibiting the same pre-definable features among a set 
of aligned sequences supplied by the user and previously obtained in other software (DOMINO
 marker selection module). </p>

<p>Furthermore, if the taxa panel has been designed in a convenient phylogenetic context 
and the user asks for highly conserved regions flanking to be included in the developed 
markers, these markers should be suitable for further surveys of variation in an extended 
set of phylogenetically related taxa, i.e. “focal taxa”. </p>

<h1>
<a id="2-installation" class="anchor" href="#2-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2. INSTALLATION</h1>

<h2>
<a id="21-supported-operating-systems" class="anchor" href="#21-supported-operating-systems" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2.1. Supported Operating Systems</h2>

<p>DOMINO is developed and maintained under GNU/Unix using ANSI/ISO C and standard 
development tools (GCC, Autotools).</p>

<p>It can be compiled and run in Linux (it has been tested in Ubuntu 14.04 LTS) and 
Mac OSX systems (version &gt; 10.10)</p>

<h2>
<a id="22-requirements" class="anchor" href="#22-requirements" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2.2. Requirements</h2>

<ul>
<li><p>Perl Programming Language version v5.2+</p></li>
<li>
<p>C++ compiler</p>

<ul>
<li>Mac OS users may need to install Xcode package via App Store</li>
<li>Linux users may need to install build essential package o g++ compiler
Ubuntu:
    sudo apt-get install build-essential 
    or 
    sudo apt-get install g++</li>
</ul>
</li>
<li>
<p>zlib compression library v1.2.3+ (SAMtools dependency) <a href="http://www.zlib.net/">http://www.zlib.net/</a></p>

<ul>
<li>Mac OS: included and installed in the distribution.</li>
<li>Linux users may need to install zlib1g-dev package
Ubuntu/Linux Mint:
    sudo apt-get install zlib1g-dev</li>
</ul>
</li>
<li>
<p>Xterm (in some ubuntu-like distributions may not be included):</p>

<ul>
<li>Mac OS: included and installed in the distribution</li>
<li>Ubuntu/Linux Mint:
sudo apt-get install xterm</li>
</ul>
</li>
</ul>

<h2>
<a id="23-installation" class="anchor" href="#23-installation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2.3. Installation</h2>

<p>Two possible installation options:</p>

<h3>
<a id="231-domino-gui-plus-command-line-version" class="anchor" href="#231-domino-gui-plus-command-line-version" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2.3.1. DOMINO GUI plus Command-line version:</h3>

<p>All necessary installation files and folders are available form the domino git directory 
or the website. The installer will guide user through the complete installation process.</p>

<p>The bin directory will contain the DOMINO perl scripts, the directories with the binaries 
of the external software included in the DOMINO pipeline and some mandatory Perl modules. 
In addition, a shortcut Desktop DOMINO icon will be generated.</p>

<h3>
<a id="232-command-line" class="anchor" href="#232-command-line" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2.3.2. Command-line:</h3>

<p>In the git directory, the user can find a pre-configured shell script (install.sh) written
to handle the installation of command-line version (please, notice that only the 
command-line version should be installed using this option).</p>

<p>In a Unix-based bash console, change the directory to fit the desired installation folder 
and run the script, with 'sh install.sh'</p>

<p>The bin directory will contain the DOMINO perl scripts, the directories with the binaries 
of the external software included in the DOMINO pipeline and some mandatory Perl modules.</p>

<h1>
<a id="2-bug-reports" class="anchor" href="#2-bug-reports" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>2. BUG REPORTS</h1>

<p>If you encounter bugs that are not listed here, please send a
report to the main authors.</p>

<pre><code>J.F.Sanchez: jfsanchezherrero@ub.edu
A.Sanchez-Gracia: elsanchez@ub.edu
J.Rozas: jrozas@ub.edu 
</code></pre>

<h1>
<a id="3-community" class="anchor" href="#3-community" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>3. COMMUNITY</h1>

<p>If you would like to see a new feature implemented, send
suggestions to the main authors (See AUTHORS). Our aim is to have an
"asked features" framework soon.</p>

<h1>
<a id="4-thanks" class="anchor" href="#4-thanks" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>4. THANKS</h1>

<p>The authors would like to thank all people who contributed (and are
still contributing) in the creation and testing of this software. </p>

<h1>
<a id="5-authors" class="anchor" href="#5-authors" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>5. AUTHORS:</h1>

<p>Cristina Frias-Lopez, Jose F. Sanchez-Herrero, Miquel A. Arnedo, Alejandro Sanchez-Gracia 
and Julio Rozas.</p>

<p>Evolutionary Genomics and Bioinformatics Group, Departament de Genètica, Microbiologia i 
Estadística and Institut de Recerca de la Biodiversitat (IRBio), Universitat de Barcelona, 
Av. Diagonal 643, Barcelona 08028, Spain</p>

<h1>
<a id="6-citation" class="anchor" href="#6-citation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>6. CITATION</h1>

<p><em>To add citation when available</em></p>

<h1>
<a id="7-copyright-and-license" class="anchor" href="#7-copyright-and-license" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>7. COPYRIGHT AND LICENSE</h1>

<p>Copyright (C) 2016 Evolutionary Genomics and Bioinformatics Group, University of Barcelona.</p>

<p>DOMINO is licensed under the GPLv3 license.  See `LICENSE' file for details. </p>

<p>This program is free software: you can redistribute it and/or modify it under the terms of 
the GNU General Public License as published by the Free Software Foundation, either version
3 of the License, or (at your option) any later version.</p>

<p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; 
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  </p>

<p>See the GNU General Public License for more details.</p>

<p>You should have received a copy of the GNU General Public License along with this program.<br>
If not, see <a href="http://www.gnu.org/licenses/">http://www.gnu.org/licenses/</a></p>
        </section>

        <aside id="sidebar">
          <a href="https://github.com/molevol-ub/DOMINO/zipball/master" class="button">
            <small>Download</small>
            .zip file
          </a>
          <a href="https://github.com/molevol-ub/DOMINO/tarball/master" class="button">
            <small>Download</small>
            .tar.gz file
          </a>

          <p class="repo-owner"><a href="https://github.com/molevol-ub/DOMINO"></a> is maintained by <a href="https://github.com/molevol-ub">molevol-ub</a>.</p>

          <p>This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the Architect theme by <a href="https://twitter.com/jasonlong">Jason Long</a>.</p>
        </aside>
      </div>
    </div>

  
  </body>
</html>
